server_test.go

```diff
@@ -575,6 +578,97 @@ func TestShutdownTCP(t *testing.T) {
 	}
 }
 
+func checkInProgressQueriesAtShutdownServer(t *testing.T, srv *Server, addr string, fin chan error, client *Client) {
```

Comment:
This test seems a bit flaky to me (from a quick read over). Is it guaranteed to always succeed?

Reply:
It is what I expect.
That is true on my local machine. Not so sure in the Travis CI. Was thinking due to timeout tuning ... 

If you have an advise on how to build better here. I would take it.

Reply:
I guess we depend of the time to write a msg on the connection.
to avoid any flakyness due to latency, we would like to be sure all msg are sent before triggering the shutdown.
But to be useful we need to wait the returned msg.
It would need to split the Exchange function and get a signal when the msg is sent.

Reply:
this tests needs to be better; it provides little context to actually go and debug when it fails. Currently it is just brute-forcing and hoping for the test. I, however, don't know a good way to test this either.

you want to know that when being shutdown you handle all sent queries... seems the that server's handle func just needs some inside in the server state to determine that any new incoming queries are still within the shutdownTimeout or something?

Reply:
there is maybe an idea here : I can hold the replies on the HandleFunc until the server is shutdown.
Doing so, all queries will be replied AFTER the shutdown ..

as long as the number of waiting replies is lower than the max workers.

Let me try that. It should be more predictable.